### PR TITLE
feat: Deprecate available_features pt 2

### PR DESCRIPTION
**Pull request by xrdt**
## Problem

Now to deprecate backend references to available_features. 

One more PR after this will be to remove the db fields. 

## Changes

👉 _Stay up-to-date with [PostHog coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review._

### CODE REVIEW

File: plugin-server/functional_tests/webhooks.test.ts [hunk 1]
```diff
@@ -120,7 +120,9 @@ test.concurrent(`webhooks: fires zapier REST webhook`, async () => {
         const distinctId = new UUIDT().toString()
         const ts = new Date()
 
-        const organizationId = await createOrganization({ available_features: '{zapier}' })
+        const organizationId = await createOrganization({
+            available_product_features: `array ['{ "key": "zapier", "name": "zapier" }'::jsonb]`,
```
Comment (xrdt) [inline plugin-server/functional_tests/webhooks.test.ts:124]: Could use some help getting this to work...

Comment (xrdt) [reply plugin-server/functional_tests/webhooks.test.ts:124]: It keeps complaining about a malformed array literal : 
```
error: malformed array literal: "array ['{ "key": "zapier", "name": "zapier" }'::jsonb]"
```

File: posthog/models/organization.py [hunk 2]
```diff
@@ -176,34 +186,37 @@ def _billing_plan_details(self) -> tuple[Optional[str], Optional[str]]:
...
 
-    def update_available_features(self) -> list[Union[AvailableFeature, str]]:
-        """Updates field `available_features`. Does not `save()`."""
+    def update_available_product_features(self) -> list[ProductFeature]:
+        """Updates field `available_product_features`. Does not `save()`."""
         if is_cloud() or self.usage:
-            # Since billing V2 we just use the available features which are updated when the billing service is called
-            return self.available_features
+            # Since billing V2 we just use the field which is updated when the billing service is called
+            return self.available_product_features  # type: ignore
```
Comment (zlwaterfield) [inline posthog/models/organization.py:193]: Why do you need the ignore now but you didn't before?

File: posthog/models/organization.py [hunk 3]
```diff
@@ -176,34 +186,37 @@ def _billing_plan_details(self) -> tuple[Optional[str], Optional[str]]:
...
 
-        self.available_features = []
+        self.available_product_features = []
 
         # Self hosted legacy license so we just sync the license features
         # Demo gets all features
         if settings.DEMO or "generate_demo_data" in sys.argv[1:2]:
-            self.available_features = License.PLANS.get(License.ENTERPRISE_PLAN, [])
+            features = License.PLANS.get(License.ENTERPRISE_PLAN, [])
+            self.available_product_features = [{"key": feature, "name": feature} for feature in features]
```
Comment (zlwaterfield) [inline posthog/models/organization.py:207]: Is the name always supposed to be the key? Or a more human-readable version of it?

File: posthog/models/user.py [hunk 4]
```diff
@@ -176,30 +176,32 @@ def teams(self):
...
-                # - the user is Admin or owner
-                teams = teams.filter(
-                    Q(access_control=False)
-                    | Q(pk__in=available_private_project_ids)
-                    | Q(organization__pk__in=organizations_where_user_is_admin)
-                )
+        org_available_product_features = Organization.objects.filter(members=self).values_list(
+            "available_product_features", flat=True
+        )
+        if org_available_product_features[0]:
```
Comment (zlwaterfield) [inline posthog/models/user.py:182]: why is this check for the first index? should it just check len > 0 ?

File: ee/session_recordings/test/test_session_recording_playlist.py [hunk 5]
```diff
@@ -79,6 +79,8 @@ def test_creates_playlist(self):
 
     def test_creates_too_many_playlists(self):
         limit = 0
+        self.organization.available_product_features = AVAILABLE_PRODUCT_FEATURES
```
Comment (zlwaterfield) [inline ee/session_recordings/test/test_session_recording_playlist.py:82]: Why is this needed now but it wasn't before?

**Review by zlwaterfield — COMMENTED** [review]
Left a few questions. I'd also recommend a review from someone on pipeline since this touches so much of the plugin server.

File: ee/session_recordings/test/test_session_recording_playlist.py (hunk 5 above)
Comment (xrdt) [reply ee/session_recordings/test/test_session_recording_playlist.py:82]: It used to be added as part of class setup. But this feature is really only relevant for this test so I figured it makes sense to pull it locally.

File: posthog/models/user.py (hunk 4 above)
Comment (xrdt) [reply posthog/models/user.py:182]: Checking for len>0 doesn't work because the org_available_product_features could be a QuerySet<None>, which has len == 1.

File: posthog/models/organization.py (hunk 2 above)
Comment (xrdt) [reply posthog/models/organization.py:193]: The typing isn't particularly friendly bc of the database field allowing nulls so I've updated this to be `return self.available_product_features or []`, which also silences the error. The problem before was that mypy wasn't happy with the possibility of the function returning None when it's signature is labelled `list[ProductFeature]`

File: posthog/models/organization.py (hunk 3 above)
Comment (xrdt) [reply posthog/models/organization.py:207]: Yeah, the name is normally a more human readable version. I can parse out the features values to look a bit nicer.

File: plugin-server/functional_tests/webhooks.test.ts (hunk 1 above)
Comment (xrdt) [reply plugin-server/functional_tests/webhooks.test.ts:124]: I also tried using jsonb_build_object

File: posthog/models/user.py (hunk 4 above)
Comment (neilkakkar) [inline posthog/models/user.py:184]: this is not ok, will fail when say this list returns an empty queryset, which while maybe a ultra rare case, our code should be defensive.

You should use .len() or .exists() in conjunction with checking the value of the first item

```suggestion
        if org_available_product_features.exists() and org_available_product_features[0]:
```

Comment (neilkakkar) [reply posthog/models/user.py:184]: or maybe even better, use `.first()`

File: plugin-server/functional_tests/webhooks.test.ts (hunk 1 above)
Comment (xrdt) [reply plugin-server/functional_tests/webhooks.test.ts:124]: That returns: 
``` 
detail: '"[" must introduce explicitly-specified array dimensions.',
```

Comment (neilkakkar) [reply plugin-server/functional_tests/webhooks.test.ts:124]: we're doing lots of wonky shenanigans with the generated SQL here (see inside createOrganization fn), easiest imo to just write the raw one in this case with the right ::jsonb coercion to make it work here, and not deal with postgres array literals at all (which don't work well with complicated jsonb conversion)

(see: https://stackoverflow.com/questions/69560635/problem-inserting-json-array-into-postgres-malformed-error )

File: ee/billing/billing_manager.py [hunk 6]
```diff
@@ -355,11 +355,6 @@ def update_org_details(self, organization: Organization, billing_status: Billing
                 org_modified = True
                 sync_org_quota_limits(organization)
 
-        available_features = data.get("available_features", None)
```
Comment (zlwaterfield) [inline ee/billing/billing_manager.py:358]: So we should be good to remove this now because the client and permission checks are all using `available_product_features`?

**Review by zlwaterfield — APPROVED** [review]

File: ee/billing/billing_manager.py (hunk 6 above)
Comment (xrdt) [reply ee/billing/billing_manager.py:358]: Did one more sweep through the codebase to confirm and ran some manual testing in the UI, including the case you had initially brought up. Everything seems good.
